compiler/src/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp

```diff
@@ -46,50 +46,76 @@ VulkanSPIRVTargetOptions getVulkanSPIRVTargetOptionsFromFlags() {
-      llvm::cl::init("unknown-unknown-unknown"));
+  };
 
-  static llvm::cl::opt<std::string> clVulkanTargetEnv(
+  static llvm::cl::list<std::string> clVulkanTargetEnvs{
       "iree-vulkan-target-env",
       llvm::cl::desc(
           "Vulkan target environment as #vk.target_env attribute assembly."),
-      llvm::cl::init(""));
+  };
 
   static llvm::cl::opt<bool> clVulkanIndirectBindings(
       "iree-vulkan-experimental-indirect-bindings",
       llvm::cl::desc("Force indirect bindings for all generated dispatches."),
       llvm::cl::init(false));
 
   VulkanSPIRVTargetOptions targetOptions;
-  targetOptions.targetEnv = clVulkanTargetEnv;
-  targetOptions.targetTriple = clVulkanTargetTriple;
+
+  int tripleCount = clVulkanTargetTriples.getNumOccurrences();
+  int envCount = clVulkanTargetEnvs.getNumOccurrences();
+  int tripleIdx = 0;
+  int envIdx = 0;
+
+  // Use the relative positions in the argument list to get the flat list of
+  // target environments.
+  SmallVector<std::string> vulkanTargetTriplesAndEnvs;
+  for (int i = 0, e = tripleCount + envCount; i < e; ++i) {
+    if (tripleIdx >= tripleCount) {
+      vulkanTargetTriplesAndEnvs.push_back(clVulkanTargetEnvs[envIdx++]);
+      continue;
+    }
+    if (envIdx >= envCount) {
+      vulkanTargetTriplesAndEnvs.push_back(clVulkanTargetTriples[tripleIdx++]);
+      continue;
+    }
+    if (clVulkanTargetTriples.getPosition(tripleIdx) >
+        clVulkanTargetEnvs.getPosition(envIdx)) {
+      vulkanTargetTriplesAndEnvs.push_back(clVulkanTargetEnvs[envIdx++]);
+    } else {
+      vulkanTargetTriplesAndEnvs.push_back(clVulkanTargetTriples[tripleIdx++]);
+    }
+  }
+  targetOptions.targetTriplesAndEnvs = vulkanTargetTriplesAndEnvs;
+
   targetOptions.indirectBindings = clVulkanIndirectBindings;
 
   return targetOptions;
 }
 
 // Returns the Vulkan target environment for conversion.
 static spirv::TargetEnvAttr
-getSPIRVTargetEnv(const std::string &vulkanTargetEnv,
-                  const std::string &vulkanTargetTriple, MLIRContext *context) {
-  if (!vulkanTargetEnv.empty()) {
-    if (auto attr = parseAttribute(vulkanTargetEnv, context)) {
-      if (auto vkTargetEnv = llvm::dyn_cast<Vulkan::TargetEnvAttr>(attr)) {
-        return convertTargetEnv(vkTargetEnv);
+getSPIRVTargetEnv(const std::string &vulkanTargetTripleOrEnv,
+                  MLIRContext *context) {
+  if (!vulkanTargetTripleOrEnv.empty()) {
+    if (vulkanTargetTripleOrEnv[0] == '#') {
+      if (auto attr = parseAttribute(vulkanTargetTripleOrEnv, context)) {
+        if (auto vkTargetEnv = llvm::dyn_cast<Vulkan::TargetEnvAttr>(attr)) {
+          return convertTargetEnv(vkTargetEnv);
+        }
       }
+      emitError(Builder(context).getUnknownLoc())
+          << "cannot parse vulkan target environment as #vk.target_env "
+             "attribute: '"
+          << vulkanTargetTripleOrEnv << "'";
+    } else {
+      return convertTargetEnv(
```

Comment:
Nit: Move this case as the first so we can do early return here and avoid one level of nesting in the previous case?